flight/PiOS/Common/pios_bmx055.c

```diff
@@ -0,0 +1,589 @@
+	uint32_t spi_slave_gyro;                    /**< The slave number (SPI) */
+	uint32_t spi_slave_accel;
+	uint32_t spi_slave_mag;
+
+	struct pios_queue *gyro_queue;
+	struct pios_queue *accel_queue;
+	struct pios_queue *mag_queue;
+
+	struct pios_thread *task_handle;
+	struct pios_semaphore *data_ready_sema;
+	volatile uint32_t interrupt_count;
+};
+
+//! Global structure for this device device
+static struct pios_bmx055_dev *bmx_dev;
+
+//! Private functions
+/**
+ * @brief Allocate a new device
+ */
+static struct pios_bmx055_dev *PIOS_BMX_Alloc(const struct pios_bmx055_cfg *cfg);
+
+/**
+ * @brief Validate the handle to the device
+ * @returns 0 for valid device or -1 otherwise
+ */
+static int32_t PIOS_BMX_Validate(struct pios_bmx055_dev *dev);
+
+static void PIOS_BMX_Task(void *parameters);
+
+/**
+ * @brief Claim the SPI bus for the communications and select this chip
+ * @return 0 if successful, -1 for invalid device, -2 if unable to claim bus
+ */
+static int32_t PIOS_BMX_ClaimBus(int slave);
+
+/**
+ * @brief Release the SPI bus for the communications and end the transaction
+ * @return 0 if successful
+ */
+static int32_t PIOS_BMX_ReleaseBus(int slave);
+
+static int32_t PIOS_BMX_ReadReg(int slave, uint8_t address, uint8_t *buffer);
+static int32_t PIOS_BMX_WriteReg(int slave, uint8_t address, uint8_t buffer);
+
+static struct pios_bmx055_dev *PIOS_BMX_Alloc(const struct pios_bmx055_cfg *cfg)
+{
+	struct pios_bmx055_dev *dev;
+
+	dev = (struct pios_bmx055_dev *)PIOS_malloc(sizeof(*bmx_dev));
+	if (!dev)
+		return NULL;
+
+	dev->magic = PIOS_BMX_DEV_MAGIC;
+
+	dev->accel_queue = PIOS_Queue_Create(PIOS_BMX_QUEUE_LEN, sizeof(struct pios_sensor_accel_data));
+	if (dev->accel_queue == NULL) {
+		PIOS_free(dev);
+		return NULL;
+	}
+
+	dev->gyro_queue = PIOS_Queue_Create(PIOS_BMX_QUEUE_LEN, sizeof(struct pios_sensor_gyro_data));
+	if (dev->gyro_queue == NULL) {
+		PIOS_Queue_Delete(dev->accel_queue);
+		PIOS_free(dev);
+		return NULL;
+	}
+
+#if 0
+	dev->mag_queue = PIOS_Queue_Create(PIOS_BMX_QUEUE_LEN, sizeof(struct pios_sensor_mag_data));
+	if (dev->mag_queue == NULL) {
+		PIOS_Queue_Delete(dev->accel_queue);
+		PIOS_Queue_Delete(dev->gyro_queue);
+		PIOS_free(dev)
+		return NULL;
+	}
+#endif
+
+	dev->data_ready_sema = PIOS_Semaphore_Create();
+	if (dev->data_ready_sema == NULL) {
+		PIOS_Queue_Delete(dev->accel_queue);
+		PIOS_Queue_Delete(dev->gyro_queue);
+		PIOS_free(dev);
+		return NULL;
+	}
+
+	return dev;
+}
+
+static int32_t PIOS_BMX_Validate(struct pios_bmx055_dev *dev)
+{
+	if (dev == NULL)
+		return -1;
+	if (dev->magic != PIOS_BMX_DEV_MAGIC)
+		return -2;
+	if (dev->spi_id == 0)
+		return -3;
+	return 0;
+}
+
+static int32_t AssertReg(int slave, uint8_t address, uint8_t expect) {
+	uint8_t c;
+
+	int32_t ret = PIOS_BMX_ReadReg(slave, address, &c);
+
+	if (ret) {
+		return ret;
+	}
+
+	if (c != expect) {
+		printf("BMX: Assertion failed: *(%d:%02x) == %02x (expect %02x)\n",
```

Comment:
So what I was planning was-- to leave this in, in the meantime.  Ultimately to work on logging infrastructure to supplant flight's debug console, and to send it to syslog(3) on simulator (this will ensure it doesn't block on console output too).

I can comment out the printf's or remove entirely in the meantime if you want, but this is placeholder for messages i want to have ultimately.